x-pack/plugins/security_solution/server/assistant/tools/esql_language_knowledge_base/esql_language_knowledge_base_tool.ts

```diff
@@ -27,11 +31,18 @@ export const ESQL_KNOWLEDGE_BASE_TOOL: AssistantTool = {
     const { chain } = params as EsqlKnowledgeBaseToolParams;
     if (chain == null) return null;
 
-    return new ChainTool({
-      name: 'ESQLKnowledgeBaseTool',
-      description:
-        'Call this for knowledge on how to build an ESQL query, or answer questions about the ES|QL query language.',
-      chain,
+    return new DynamicTool({
+      name: toolDetails.name,
+      description: toolDetails.description,
+      func: async (input, _, cbManager) => {
+        const result = await chain.invoke(
+          {
+            query: getEsqlPrompt(input),
```

Comment:
I'm wondering if there may be an unintended side effect here, because the instructions:

```
Only return valid ES|QL queries as described above. Do not add any additional text to describe your output.
```

appear to be included in the ELSER query, as illustrated in [this run](https://smith.langchain.com/public/11331f5d-5fd8-4a24-9a42-6699bf6d2c35/r/2ed50f35-f114-431c-b9d8-2d9e1b771896) linked in the PR description:

![vector_store_retriever](https://github.com/elastic/kibana/assets/4459398/b5209aaf-ab05-4684-bde3-43d4901e91a8)

or illustrated by adding the following `console.log`:

```typescript
    try {
      console.log('mSearchQueryBody:', JSON.stringify(mSearchQueryBody, null, 2));

      // execute both queries via a single multi-search request:
      const result = await this.esClient.msearch<MsearchResponse>(mSearchQueryBody);
```

to `x-pack/plugins/elastic_assistant/server/lib/langchain/elasticsearch_store/elasticsearch_store.ts`,

which produces the following Elasticsearch query:

```json
{
  "body": [
    {
      "index": ".kibana-elastic-ai-assistant-kb"
    },
    {
      "query": {
        "bool": {
          "must_not": [
            {
              "term": {
                "metadata.kbResource": "esql"
              }
            },
            {
              "term": {
                "metadata.required": true
              }
            }
          ],
          "must": [
            {
              "text_expansion": {
                "vector.tokens": {
                  "model_id": ".elser_model_2",
                  "model_text": "\nSELECT host.name, percentile(bytes, 99) AS p99, percentile(bytes, 95) AS p95, percentile(bytes, 75) AS p75 FROM \"logstash-*\" GROUP BY host.name ORDER BY COUNT(*) DESC LIMIT 10\n\nOnly return valid ES|QL queries as described above. Do not add any additional text to describe your output.\n "
                }
              }
            }
          ]
        }
      },
      "size": 10
    },
    {
      "index": ".kibana-elastic-ai-assistant-kb"
    },
    {
      "query": {
        "bool": {
          "must": [
            {
              "term": {
                "metadata.kbResource": "esql"
              }
            },
            {
              "term": {
                "metadata.required": true
              }
            }
          ]
        }
      },
      "size": 10000
    }
  ]
}
```

Reply:
I wonder if there is another argument we can pass besides `query`? Clicking through the code, this type is `Record<string, any>` so I am having a hard time determining what properties actually matter. Have you found better docs on the invoke function arguments?

Reply:
> I wonder if there is another argument we can pass besides `query`? Clicking through the code, this type is `Record<string, any>` so I am having a hard time determining what properties actually matter. Have you found better docs on the invoke function arguments?

I'm not seeing docs apart from the [RetrievalQAChain](https://api.js.langchain.com/classes/langchain_chains.RetrievalQAChain.html)'s documentation of [invoke](https://api.js.langchain.com/classes/langchain_chains.RetrievalQAChain.html#invoke), which doesn't appear to document any additional args.

Reply:
Spoke with Andrew over slack and we decided to [move the line from the query argument to the description](https://github.com/elastic/kibana/pull/181088/commits/c8e80c6933103ee92a6dbcd46ba6b5ea1ff10d77). This passes 10/10 times with Sonnet, consistently giving 1 of 2 similar (accurate) responses.